pkg/registry/apis/provisioning/register.go

```diff
@@ -0,0 +1,160 @@
+		!features.IsEnabledGlobally(featuremgmt.FlagGrafanaAPIServerWithExperimentalAPIs) {
+		return nil, nil // skip registration unless opting into experimental apis OR the feature specifically
+	}
+
+	builder := NewAPIBuilder()
+	apiregistration.RegisterAPI(builder)
+	return builder, nil
+}
+
+func (b *APIBuilder) GetAuthorizer() authorizer.Authorizer {
+	return authorizer.AuthorizerFunc(
+		func(ctx context.Context, a authorizer.Attributes) (authorizer.Decision, string, error) {
+			// TODO: Implement a webhook authoriser somehow.
+
+			// fallback to the standard authorizer
+			return authorizer.DecisionNoOpinion, "", nil
+		})
+}
+
+func (b *APIBuilder) GetGroupVersion() schema.GroupVersion {
+	return provisioning.SchemeGroupVersion
+}
+
+func (b *APIBuilder) InstallSchema(scheme *runtime.Scheme) error {
+	err := provisioning.AddToScheme(scheme)
+	if err != nil {
+		return err
+	}
+
+	// This is required for --server-side apply
+	err = provisioning.AddKnownTypes(provisioning.InternalGroupVersion, scheme)
+	if err != nil {
+		return err
+	}
+
+	metav1.AddToGroupVersion(scheme, provisioning.SchemeGroupVersion)
+	// Only 1 version (for now?)
+	return scheme.SetVersionPriority(provisioning.SchemeGroupVersion)
+}
+
+func (b *APIBuilder) UpdateAPIGroupInfo(apiGroupInfo *genericapiserver.APIGroupInfo, opts builder.APIGroupOptions) error {
+	repositoryStorage, err := grafanaregistry.NewRegistryStore(opts.Scheme, provisioning.RepositoryResourceInfo, opts.OptsGetter)
+	if err != nil {
+		return fmt.Errorf("failed to create repository storage: %w", err)
+	}
+
+	repositoryStatusStorage := grafanaregistry.NewRegistryStatusStore(opts.Scheme, repositoryStorage)
+
+	storage := map[string]rest.Storage{}
+	storage[provisioning.RepositoryResourceInfo.StoragePath()] = repositoryStorage
+	storage[provisioning.RepositoryResourceInfo.StoragePath("status")] = repositoryStatusStorage
+	apiGroupInfo.VersionedResourcesStorageMap[provisioning.VERSION] = storage
+	return nil
+}
+
+func (b *APIBuilder) Mutate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	obj := a.GetObject()
+
+	if obj == nil || a.GetOperation() == admission.Connect {
+		return nil // This is normal for sub-resource
+	}
+
+	r, ok := obj.(*provisioning.Repository)
+	if !ok {
+		return fmt.Errorf("expected repository configuration")
+	}
+
+	// TODO: Do something based on the resource we got.
+	_ = r
+
+	return nil
+}
+
+func (b *APIBuilder) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	obj := a.GetObject()
+	if obj == nil || a.GetOperation() == admission.Connect {
+		return nil // This is normal for sub-resource
+	}
+
+	var list field.ErrorList
+	// TODO: Fill the list with validation errors.
+
+	if len(list) > 0 {
+		return apierrors.NewInvalid(
+			provisioning.RepositoryResourceInfo.GroupVersionKind().GroupKind(),
+			a.GetName(), list)
+	}
+	return nil
```

Comment:
🤔  should we fail trying to write anything at this sate?  so we do not accidentallly save files before we have any real controls

Reply:
I don't think it really matters. The feature is feature-flagged and not intended for any use whatsoever except via the feature branch 😅 

Validation is probably one of the absolute first things we'll want to migrate as well